llama/model_cn_comment.py

```diff
@@ -86,50 +86,69 @@ def forward(self, x):
 
 
 def precompute_freqs_cis(dim: int, end: int, theta: float = 10000.0):
-    """Precompute the frequency tensor for complex exponentials (cis) with given dimensions.
+    """根据指定维度预计算复指数 (cis) 的频率张量.
 
-    This function calculates a frequency tensor with complex exponentials using the given dimension
-    'dim' and the end index 'end'. The 'theta' parameter scales the frequencies.
-    The returned tensor contains complex values in complex64 data type.
+    该函数根据给定的维度 dim 和结束索引 end, 计算一个复指数的频率张量.
+    theta 参数用于缩放频率.
+    返回的张量包含复数值，数据类型为 complex64.
 
     Args:
-        dim (int): Dimension of the frequency tensor.
-        end (int): End index for precomputing frequencies.
-        theta (float, optional): Scaling factor for frequency computation. Defaults to 10000.0.
+        dim (int): 频率张量的维度, 也就是 self-attention 中 query 和 key 向量的维度.
+        end (int): 预计算频率的结束索引, 也就是模型支持的最长序列长度.
+        theta (float, optional): 频率计算的缩放因子, 默认为10000.0.
 
     Returns:
-        torch.Tensor: Precomputed frequency tensor with complex exponentials.
+        torch.Tensor: 预计算的复指数的频率张量. Shape: (end, dim / 2).
     """
+    # 根据 theta 和 dim 参数计算用于 RoPE 不同维度的旋转角度的基.
+    # 即计算 θ_i = theta^{-2i / dim}, i ∈ [0, dim / 2).
+    # [Shape] freqs: (dim / 2,)
     freqs = 1.0 / (
         theta ** (torch.arange(0, dim, 2)[: (dim // 2)].float() / dim)
     )
-    t = torch.arange(end, device=freqs.device)  # type: ignore
-    freqs = torch.outer(t, freqs).float()  # type: ignore
+    # 根据结束索引 end 生成用于采样所有位置旋转角度的 t.
+    # [Shape] t: (end,)
+    t = torch.arange(end, device=freqs.device)
+    # 通过向量外积的操作采样所有位置对应维度的旋转角度.
+    # 即对于 t 中的每一个位置 m 和 freqs 中的每一组维度对应的角度基 θ_i,
+    # 采样出对应位置和维度的旋转角度 mθ_i.
+    # [Shape] freqs: (end, dim / 2)
+    freqs = torch.outer(t, freqs).float()
+    # 将得到的所有位置对应维度的旋转角度转化为复指数频率.
+    # 即对于每个旋转角度 θ, 计算对应的频率 e^{iθ}.
+    # 其操作参考：docs/CN/RoPE 中关于欧拉公式的描述.
+    # [Shape] freqs_cis: (end, dim / 2)
+    # [Type] freqs_cis: complex64
     freqs_cis = torch.polar(torch.ones_like(freqs), freqs)  # complex64
     return freqs_cis
 
 
 def reshape_for_broadcast(freqs_cis: torch.Tensor, x: torch.Tensor):
-    """Reshape frequency tensor for broadcasting it with another tensor.
+    """Reshape 频率张量以便与另一个张量进行广播.
 
-    This function reshapes the frequency tensor to have the same shape as the target tensor 'x'
-    for the purpose of broadcasting the frequency tensor during element-wise operations.
+    此函数将频率张量 Reshape 成与目标张量 x 相同的形状, 以便在逐元素操作中广播频率张量.
 
     Args:
-        freqs_cis (torch.Tensor): Frequency tensor to be reshaped.
-        x (torch.Tensor): Target tensor for broadcasting compatibility.
+        freqs_cis (torch.Tensor): 需要 reshape 的频率张量. Shape: (end, dim / 2).
+        x (torch.Tensor): 目标张量，用于确保广播兼容性. Shape: (batch_size, n, n_heads, dim / 2).
+            注意, 该张量为复数张量.
 
     Returns:
-        torch.Tensor: Reshaped frequency tensor.
+        torch.Tensor: 已重塑的频率张量.
```

Comment:
Keep consistent, how about "Reshape 后的频率张量"?

Reply:
Add the shape of the returned tensor will be better.
```suggestion
        torch.Tensor: Reshape 后的频率张量. Shape: (...)
```